tracing-attributes/src/lib.rs

```diff
@@ -262,14 +273,40 @@ pub fn instrument(args: TokenStream, item: TokenStream) -> TokenStream {
     // If the function is an `async fn`, this will wrap it in an async block,
     // which is `instrument`ed using `tracing-futures`. Otherwise, this will
     // enter the span and then perform the rest of the body.
+    // If `err` is in args, instrument any resulting `Err`s.
     let body = if asyncness.is_some() {
-        quote_spanned! {block.span()=>
-            tracing_futures::Instrument::instrument(
-                async move { #block },
-                __tracing_attr_span
-            )
-                .await
+        if instrument_err(&args) {
+            quote_spanned! {block.span()=>
+                tracing_futures::Instrument::instrument(async move {
+                    match async move { #block }.await {
+                        Ok(x) => Ok(x),
+                        Err(e) => {
+                            tracing::error!("{}", e);
```

Comment:
This relies on the error implementing `fmt::Display`, and records it using its display implementation. We _might_ want to use the [`tracing::field::Value` implementation for `dyn Error + 'static`](https://docs.rs/tracing/0.1.13/tracing/field/trait.Value.html#foreign-impls) instead &mdash; that way, the subscriber could choose to use the error's `Debug` implementation, downcast it, or traverse its `source` chain, as well as recording it using `Display`.

In that case, this will require the error type to implement the `Error` trait, which should probably be documented. But, since we are already requiring a `Display` impl, the traits required for the error type should probably be documented either way.

Reply:
Thought about this, but in my particular use case I'm actually using [`anyhow::Error`](https://docs.rs/anyhow/1.0.27/anyhow/struct.Error.html), which doesn't impl Error directly but provides `impl AsRef<dyn std::error::Error + 'static> for Error`.

Prior to making these changes, I had a macro that did:

```rust
let error: &(dyn std::error::Error + 'static) = $error.as_ref();
tracing::error!(error = error);
```

but I don't know if that's generally applicable to other types that `impl Error`.

Reply:
> Prior to making these changes, I had a macro that did:
> 
> ```rust
> let error: &(dyn std::error::Error + 'static) = $error.as_ref();
> tracing::error!(error = error);
> ```
> 
> but I don't know if that's generally applicable to other types that `impl Error`.

I don't _think_ this will work in the general case, since the impl of `AsRef` for `&'_ T` is `AsRef<U> where T: AsRef<U>`, so I don't believe that all types implementing `Error` will have `AsRef<dyn Error + 'static>` impls. 

It looks like `anyhow::Error` also has a `Deref<Target = std::error::Error + Send + Sync + 'static>` impl. It's possible we could do `*&error` or something, which might work for all `Error`s. But...I'm fine with using the display formatter if we can't come up with something that works for all error types.

One note, though: if we're going to use the display formatter, I think it's better to record the error's display impl as a field named `error` (or something), rather than using format args, which will record the error as a field named `message`. So, we could write
```rust
tracing::error!(error = %e);
```